R/matching.R

```diff
@@ -112,44 +132,30 @@ closest <- function(x, table, tolerance = Inf, ppm = 0,
         stop("'tolerance' hat to be of length 1 or equal to 'length(table)'")
 
     tolerance <- tolerance + ppm(table, ppm) + sqrt(.Machine$double.eps)
-    duplicates <- match.arg(duplicates)
-
-    lIdx <- findInterval(x, table, rightmost.closed = FALSE, all.inside = TRUE)
-    rIdx <- lIdx + 1L
-
-    lIdx[lIdx == 0L] <- 1L
-    rIdx[rIdx > ntable] <- ntable
-
-    lDiff <- abs(table[lIdx] - x)
-    rDiff <- abs(table[rIdx] - x)
 
-    lDiff[lDiff > tolerance[lIdx]] <- Inf
-    rDiff[rDiff > tolerance[rIdx]] <- Inf
-
-    d <- which(lDiff > rDiff)
-
-    lIdx[d] <- rIdx[d]
-
-    ## no match at all
-    lIdx[is.infinite(lDiff) & is.infinite(rDiff)] <- NA_integer_
-
-    ## duplicated matches
-    if (duplicates == "remove") {
-        lIdx[duplicated(lIdx) | duplicated(lIdx, fromLast = TRUE)] <-
-            NA_integer_
-        lIdx[is.finite(lDiff) & is.finite(rDiff)] <- NA_integer_
-    } else if (duplicates == "closest") {
-        ## lIdx could be updated
-        o <- order(abs(table[lIdx] - x))
-        m <- lIdx[o]
-        m[duplicated(m)] <- NA_integer_
-        lIdx[o] <- m
-    }
-
-    if (!identical(nomatch, NA_integer_))
-        lIdx[is.na(lIdx)] <- nomatch
+    duplicates <- match.arg(duplicates)
 
-    as.integer(lIdx)
+    if (duplicates == "keep")
```

Comment:
what if you used `.Call(paste0("C_closest_dup_", duplicates), ...)`? We could get rid of all these `if` `else` statements - and maybe it would be even slightly faster?

Reply:
My initial idea was that the R compiler could better work with an `if`/`else` than with a `paste` but that seems not to be the case. Nevertheless `paste0` is slower:
```r
devtools::load_all()
library("compiler")

set.seed(42)
x <- sort(sample(1000, 100) + runif(100))
y <- sort(sample(1000, 700))

cmpIfElse <- cmpfun(cIfElse)
cmpPaste <- cmpfun(cPaste)

library("microbenchmark")
microbenchmark(cIfElse(x, y), cPaste(x, y), cmpIfElse(x, y), cmpPaste(x, y))
# Unit: microseconds
#             expr    min      lq     mean  median      uq     max neval
#    cIfElse(x, y) 25.799 26.3710 27.54267 26.7995 28.5360  41.525   100
#     cPaste(x, y) 29.878 30.4775 31.42049 30.8500 32.6400  34.705   100
#  cmpIfElse(x, y) 24.651 25.6600 28.64918 26.0425 27.7225 217.980   100
#   cmpPaste(x, y) 29.121 29.9050 32.22717 30.2515 31.0400 171.343   100
#  cld
#   a 
#   ab
#   ab
#    b
``

Reply:
well, then let's keep the `if/else`.